docs/other-topics/hooks.md

```diff
@@ -2,404 +2,253 @@
-* `sequelize.beforeDisconnect(callback)`
-  * The callback has the form `async (connection) => /* ... */`
-* `sequelize.afterDisconnect(callback)`
-  * The callback has the form `async (connection) => /* ... */`
-
-These hooks can be useful if you need to asynchronously obtain database credentials, or need to directly access the low-level database connection after it has been created.
-
-For example, we can asynchronously obtain a database password from a rotating token store, and mutate Sequelize's configuration object with the new credentials:
-
-```js
-sequelize.beforeConnect(async (config) => {
-  config.password = await getAuthToken();
-});
-```
-
-These hooks may *only* be declared as a permanent global hook, as the connection pool is shared by all models.
-
-## Instance hooks
-
-The following hooks will emit whenever you're editing a single object:
-
-* `beforeValidate`
-* `afterValidate` / `validationFailed`
-* `beforeCreate` / `beforeUpdate` / `beforeSave` / `beforeDestroy`
-* `afterCreate` / `afterUpdate` / `afterSave` / `afterDestroy`
-
-```js
-User.beforeCreate(user => {
-  if (user.accessLevel > 10 && user.username !== "Boss") {
-    throw new Error("You can't grant this user an access level above 10!");
-  }
-});
-```
-
-The following example will throw an error:
-
-```js
-try {
-  await User.create({ username: 'Not a Boss', accessLevel: 20 });
-} catch (error) {
-  console.log(error); // You can't grant this user an access level above 10!
 };
-```
 
-The following example will be successful:
+Book.hooks.addListener('afterCreate', 'yourHookIdentifier', myListener);
 
-```js
-const user = await User.create({ username: 'Boss', accessLevel: 20 });
-console.log(user); // user object with username 'Boss' and accessLevel of 20
+// Both of these will remove the hook:
+// success-next-line
+Book.hooks.removeListener('afterCreate', myListener);
+// success-next-line
+Book.hooks.removeListener('afterCreate', 'yourHookIdentifier');
 ```
 
-### Model hooks
-
-Sometimes you'll be editing more than one record at a time by using methods like `bulkCreate`, `update` and `destroy`. The following hooks will emit whenever you're using one of those methods:
+:::caution
 
-* `YourModel.beforeBulkCreate(callback)`
-  * The callback has the form `(instances, options) => /* ... */`
-* `YourModel.beforeBulkUpdate(callback)`
-  * The callback has the form `(options) => /* ... */`
-* `YourModel.beforeBulkDestroy(callback)`
-  * The callback has the form `(options) => /* ... */`
-* `YourModel.afterBulkCreate(callback)`
-  * The callback has the form `(instances, options) => /* ... */`
-* `YourModel.afterBulkUpdate(callback)`
-  * The callback has the form `(options) => /* ... */`
-* `YourModel.afterBulkDestroy(callback)`
-  * The callback has the form `(options) => /* ... */`
+Hooks added by decorators cannot be removed using the callback instance, but can still be removed if they are named:
```

Comment:
Not for this PR, but will this be possible when we add decorators to sequelize itself? Or would this not be worth is since we recommend against it?

Reply:
It can't because we're binding the function to the model. We could keep a weakmap of old function to new function to preserve the functionality but I do think it wouldn't be worth it. We could even consider not allowing to remove hooks added by decorators at all

Reply:
Let's keep it like this for now